src/lib/bitcoin/btc-base-api.js

```diff
@@ -0,0 +1,212 @@
+
+      let txid = bitcoin.crypto.sha256(Buffer.from(hex, 'hex'))
+      txid = bitcoin.crypto.sha256(Buffer.from(txid))
+      txid = txid.toString('hex').match(/.{2}/g).reverse().join('')
+
+      return { hex, txid }
+    })
+  }
+
+  /**
+   * Broadcasts the specified transaction to the network.
+   * @abstract
+   * @param {string} txHex raw transaction as a HEX literal
+   */
+  sendTransaction (txHex) {
+    return Promise.resolve('')
+  }
+
+  /**
+   * Retrieves transaction details
+   * @abstract
+   * @param {*} txid transaction ID
+   * @returns {Promise<object>}
+   */
+  getTransaction (txid) {
+    return Promise.resolve(null)
+  }
+
+  /**
+   * Retrieves transactions for the specified address
+   * @abstract
+   * @param {any} options crypto-specific options
+   * @returns {Promise<{hasMore: boolean, items: Array}>}
+   */
+  getTransactions (options) {
+    return Promise.resolve({ hasMore: false, items: [] })
+  }
+
+  /**
+   * Retrieves unspents (UTXO)
+   * @abstract
+   * @returns {Promise<Array<{txid: string, vout: number, amount: number}>>}
+   */
+  _getUnspents () {
+    return Promise.resolve([])
+  }
+
+  /**
+   * Creates a raw DOGE transaction as a hex string.
+   * @param {string} address target address
+   * @param {number} amount amount to send
+   * @param {Array<{txid: string, amount: number, vout: number}>} unspents unspent transaction to use as inputs
+   * @returns {string}
+   */
+  _buildTransaction (address, amount, unspents) {
+    amount = new BigNumber(amount).times(this.multiplier).toNumber()
+
+    const txb = new bitcoin.TransactionBuilder(this._network)
+    txb.setVersion(1)
+
+    const target = amount + new BigNumber(this.getFee()).times(this.multiplier).toNumber()
+    let transferAmount = 0
+    let inputs = 0
+
+    unspents.forEach(tx => {
+      const amt = Math.floor(tx.amount)
+      if (transferAmount < target) {
+        txb.addInput(tx.txid, tx.vout)
+        transferAmount += amt
+        inputs++
+      }
+    })
+
+    txb.addOutput(address, amount)
+    txb.addOutput(this._address, transferAmount - target)
+
+    for (let i = 0; i < inputs; ++i) {
+      txb.sign(i, this._keyPair)
+    }
+
+    return txb.build().toHex()
+  }
+
+  /** Picks a client for a random API endpoint */
+  _getClient () {
+    const url = getEnpointUrl(this._crypto)
+    if (!this._clients[url]) {
+      this._clients[url] = axios.create({
+        baseURL: url
+      })
+    }
+    return this._clients[url]
+  }
+
+  _mapTransaction (tx) {
+    const senders = getUnique(tx.vin.map(x => x.addr))
+
+    const direction = senders.includes(this._address) ? 'from' : 'to'
+
+    const recipients = getUnique(tx.vout.reduce((list, out) => {
+      list.push(...out.scriptPubKey.addresses)
+      return list
+    }, []))
+
+    if (direction === 'from') {
+      // Disregard our address for the outgoing transaction
+      const idx = recipients.indexOf(this._address)
+      if (idx >= 0) recipients.splice(idx, 1)
```

Comment:
Do this only if `recipients.length > 1` (see https://github.com/Adamant-im/adamant-im/pull/237/files)